core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java

```diff
@@ -91,49 +98,111 @@ public ListenableFuture<Void> execute(
+            }
+
+            List<String> parentPath = statement.getColumn().getName().getOriginalParts().subList(0, statement.getColumn().getName().getOriginalParts().size() - 1).stream()
+                    .map(identifier -> identifier.getValue().toLowerCase(ENGLISH))
+                    .collect(toImmutableList());
+            List<String> fieldPath = statement.getColumn().getName().getOriginalParts().subList(1, statement.getColumn().getName().getOriginalParts().size()).stream()
+                    .map(Identifier::getValue)
+                    .collect(toImmutableList());
+
+            ColumnMetadata columnMetadata = plannerContext.getMetadata().getColumnMetadata(session, tableHandle, columnHandles.get(columnName.getValue().toLowerCase(ENGLISH)));
+            Type currentType = columnMetadata.getType();
+            int fieldDepth = 0;
+            for (String fieldName : fieldPath) {
+                fieldDepth++;
+                if (currentType instanceof RowType rowType) {
+                    List<RowType.Field> candidates = rowType.getFields().stream()
+                            // case-insensitive match
+                            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equalsIgnoreCase(fieldName))
+                            .collect(toImmutableList());
+                    if (fieldDepth == fieldPath.size()) {
+                        if (!candidates.isEmpty()) {
+                            if (statement.isColumnNotExists()) {
+                                return immediateVoidFuture();
+                            }
+                            throw semanticException(COLUMN_ALREADY_EXISTS, statement, "Field '%s' already exists", fieldName);
+                        }
+                        break;
+                    }
+                    if (candidates.size() > 1) {
+                        throw semanticException(AMBIGUOUS_NAME, statement, "Field path %s within %s is ambiguous", fieldPath, columnMetadata.getType());
+                    }
+                    if (candidates.size() == 1) {
+                        RowType.Field rowField = getOnlyElement(candidates);
+                        currentType = rowField.getType();
+                        continue;
+                    }
+                }
+                // TODO: Support array and map types
+                throw semanticException(
+                        COLUMN_NOT_FOUND,
+                        statement,
+                        "Cannot resolve field '%s' within %s type when adding %s in %s",
+                        fieldName,
+                        currentType,
+                        String.join(".", fieldPath),
+                        columnMetadata.getType());
+            }
+
+            String fieldName = getLast(statement.getColumn().getName().getParts());
+            plannerContext.getMetadata().addField(session, tableHandle, parentPath, fieldName, type, statement.isColumnNotExists());
```

Comment:
This is hard to follow due to the many levels of nesting, break/continue, etc.

You can improve it by:
1. Extracting the field resolution to a separate method:

```java
private static List<RowType.Field> getCandidates(RowType rowType, String fieldName)
{
    List<RowType.Field> candidates = rowType.getFields().stream()
            // case-insensitive match
            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equalsIgnoreCase(fieldName))
            .collect(toImmutableList());

    return candidates;
}
```

2. Restructure the loop as follows:

```java
for (int i = 0; i < fieldPath.size() - 1; i++) {
    String fieldName = fieldPath.get(i);

    if (currentType instanceof RowType rowType) {
        List<RowType.Field> candidates = getCandidates(rowType, fieldName);

        if (candidates.isEmpty()) {
            throw semanticException( ... field not found ...);
        }
        if (candidates.size() > 1) {
            throw semanticException(... name is ambiguous ...);
        }
        currentType = getOnlyElement(candidates).getType();
    }
    else {
        throw semanticException( ... field is not of type ROW ...);
    }
}

String fieldName = getLast(statement.getColumn().getName().getParts());
List<RowType.Field> candidates = getCandidates((RowType) currentType, fieldName);

if (!candidates.isEmpty()) {
    if (statement.isColumnNotExists()) {
        return immediateVoidFuture();
    }
    throw semanticException(COLUMN_ALREADY_EXISTS, statement, "Field '%s' already exists", fieldName);
}

...
```

Reply:
Thanks for your suggestion. Updated.